zalando.org_routegroups.yaml

```diff
@@ -214,6 +214,27 @@ spec:
                   type: object
                 minItems: 1
                 type: array
+              tls:
+                description: TLS defines which Kubernetes secret will be used to terminate
+                  the connection based on the matching hostnames
+                items:
+                  properties:
+                    hosts:
+                      description: Host specifies the list of hosts included in the
+                        TLS secret. The values in this list must match the name/s
+                        used in the tlsSecret.
+                      items:
+                        pattern: "^[a-z0-9]([-a-z0-9]*[a-z0-9])?([.][a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
```

Comment:
Where does this pattern come from?

Reply:
Yeah would be good to know. It looks also a bit wrong.
No *.example possible but you can have .foo.example

Reply:
This was generated automatically from `make zalando.org_routegroups.yaml`

Reply:
Ok, I was wondering where it comes from because I did not see any pattern defined in types.go
I turns our we have a perl snippet that applies to this field as well by luck:
https://github.com/szuecs/routegroup-client/blob/98ae2ff4e8c830d90f13f2a9fd626034eb057a39/Makefile#L33-L37

Reply:
@szuecs I think its will handle domains properly, see https://regex101.com/r/DjRVGZ/1

Reply:
> we have a perl snippet that applies to this field

Its a hack anyways and it works so I guess its fine.

Reply:
I looked again at https://github.com/kubernetes-sigs/controller-tools/issues/342 which is still not fixed. One workaround mentioned is to introduce type alias but that would be a breaking change unfortunately: 

```diff
diff --git a/Makefile b/Makefile
index 619a1a3..5db97a0 100644
--- a/Makefile
+++ b/Makefile
@@ -32,9 +32,6 @@ $(GENERATED): $(CRD_TYPE_SOURCE)
 
 $(GENERATED_CRD): go.mod $(GENERATED)
        go run sigs.k8s.io/controller-tools/cmd/controller-gen crd:crdVersions=v1 paths=./apis/... output:crd:dir=.
-       # workaround to add pattern to array items. Not supported by controller-gen
-       # ref: https://github.com/kubernetes-sigs/controller-tools/issues/342
-       perl -i -p0e 's|(\s*)(hosts:.*?items:)|$$1$$2$$1    pattern: "^[a-z0-9]([-a-z0-9]*[a-z0-9])?([.][a-z0-9]([-a-z0-9]*[a-z0-9])?)*\$$"|sg' $(GENERATED_CRD)
 
 build.local: $(LOCAL_BINARIES) $(GENERATED_CRD)
 build.linux: $(LINUX_BINARIES) $(GENERATED_CRD)
diff --git a/apis/zalando.org/v1/types.go b/apis/zalando.org/v1/types.go
index a45de6a..a1e0883 100644
--- a/apis/zalando.org/v1/types.go
+++ b/apis/zalando.org/v1/types.go
@@ -39,10 +39,13 @@ type RouteGroupList struct {
        Items []RouteGroup `json:"items"`
 }
 
+// +kubebuilder:validation:Pattern="^[a-z0-9]([-a-z0-9]*[a-z0-9])?([.][a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
+type Hostname string
+
 // +k8s:deepcopy-gen=true
 type RouteGroupSpec struct {
        // List of hostnames for the RouteGroup
-       Hosts []string `json:"hosts,omitempty"`
+       Hosts []Hostname `json:"hosts,omitempty"`
        // List of backends that can be referenced in the routes
        Backends []RouteGroupBackend `json:"backends"`
        // DefaultBackends is a list of default backends defined if no explicit
diff --git a/apis/zalando.org/v1/zz_generated.deepcopy.go b/apis/zalando.org/v1/zz_generated.deepcopy.go
index eea8969..d9ece11 100644
--- a/apis/zalando.org/v1/zz_generated.deepcopy.go
+++ b/apis/zalando.org/v1/zz_generated.deepcopy.go
@@ -187,7 +187,7 @@ func (in *RouteGroupSpec) DeepCopyInto(out *RouteGroupSpec) {
        *out = *in
        if in.Hosts != nil {
                in, out := &in.Hosts, &out.Hosts
-               *out = make([]string, len(*in))
+               *out = make([]Hostname, len(*in))
                copy(*out, *in)
        }
        if in.Backends != nil {
diff --git a/cli/rg-client-test/main.go b/cli/rg-client-test/main.go
index 60802d4..4c3394d 100644
--- a/cli/rg-client-test/main.go
+++ b/cli/rg-client-test/main.go
@@ -41,7 +41,7 @@ func main() {
        // test create
        name := "myrg"
        namespace := "default"
-       hostname := "myrg.teapot-e2e.zalan.do"
+       hostname := rgv1.Hostname("myrg.teapot-e2e.zalan.do")
        port := 83
        newRg := &rgv1.RouteGroup{
                ObjectMeta: metav1.ObjectMeta{
@@ -49,7 +49,7 @@ func main() {
                        Namespace: namespace,
                },
                Spec: rgv1.RouteGroupSpec{
-                       Hosts: []string{hostname},
+                       Hosts: []rgv1.Hostname{hostname},
                        Backends: []rgv1.RouteGroupBackend{
                                {
                                        Name:        name,
diff --git a/zalando.org_routegroups.yaml b/zalando.org_routegroups.yaml
index bd78b61..55cd9b3 100644
--- a/zalando.org_routegroups.yaml
+++ b/zalando.org_routegroups.yaml
@@ -143,7 +143,7 @@ spec:
               hosts:
                 description: List of hostnames for the RouteGroup
                 items:
-                  pattern: "^[a-z0-9]([-a-z0-9]*[a-z0-9])?([.][a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
+                  pattern: ^[a-z0-9]([-a-z0-9]*[a-z0-9])?([.][a-z0-9]([-a-z0-9]*[a-z0-9])?)*$
                   type: string
                 type: array
               routes:
```

Reply:
The implementation will check that value is also in spec.hosts which has this pattern.
We can rethink or relax pattern later but for now I think its fine. Or what do you suggest?

Reply:
ok